src/charm.py

```diff
@@ -738,10 +843,108 @@ def _get_initialization_secret_from_peer_relation(self) -> Tuple[str, List[str]]
         except (TypeError, SecretNotFoundError, AttributeError):
             raise PeerSecretError(secret_name=VAULT_INITIALIZATION_SECRET_LABEL)
 
+    def _create_s3_session(self, s3_parameters: dict) -> Optional[boto3.session.Session]:
```

Comment:
It feels like all the `S3` related actions could be in a class of their own

```
class S3():
    def __init__(self, access_key: str, secret_key: str, region: str):
        self.session = self._create_session()
    
    def _create_session():
        ...

    def create_bucket():
        ...

    def upload_content():
        ...

```

This should simplify the code and tests